### PR TITLE
Make federation of unlisted videos an instance-level server preference

### PR DESCRIPTION
**Pull request by Tak**
As discussed on #944

The previous behavior is the default for the preference.
I've added tests for:
- the preference itself
- actual video federation based on the preference
- accessing unlisted videos on the origin server

### CODE REVIEW

File: config/default.yaml [hunk 1]
```diff
@@ -325,6 +325,7 @@ instance:
   # By default, "do_not_list" or "blur" or "display" NSFW videos
   # Could be overridden per user with a setting
   default_nsfw_policy: 'do_not_list'
+  federate_unlisted_videos: true
```
Comment (Chocobozzz) [inline config/default.yaml:328]: Please put this in a dedicated key, above https://github.com/Chocobozzz/PeerTube/blob/develop/config/production.yaml.example#L183:

```
federation:
  videos:
    federated_unlisted: true
```

File: server/helpers/express-utils.ts [hunk 2]
```diff
@@ -28,6 +29,19 @@ function buildNSFWFilter (res?: express.Response, paramNSFW?: string) {
   return null
 }
 
+function isPrivacyForFederation(privacy: VideoPrivacy) {
```
Comment (Chocobozzz) [inline server/helpers/express-utils.ts:32]: Why did you move these functions in this file? `VideoModel` or `helpers/videos.ts` seems more appropriate

File: config/default.yaml (hunk 1 above)
Comment (Chocobozzz) [reply config/default.yaml:328]: If you want admins to be able to update this setting using the web interface, then put this at the end of the produdction.yaml file. Don't forget to also update the default.yaml file

File: client/src/app/+admin/config/edit-custom-config/edit-custom-config.component.ts [hunk 3]
```diff
@@ -137,6 +137,11 @@ export class EditCustomConfigComponent extends FormReactive implements OnInit, A
           css: null
         }
       },
+      federation: {
+        videos: {
+          federateUnlisted: true
```
Comment (Chocobozzz) [inline client/src/app/+admin/config/edit-custom-config/edit-custom-config.component.ts:142]: Do you want admins to update this setting on the web interface? If no, you can remove this

File: client/src/app/core/server/server.service.ts [hunk 4]
```diff
@@ -58,6 +58,11 @@ export class ServerService {
       registeredExternalAuths: [],
       registeredIdAndPassAuths: []
     },
+    federation: {
+      videos: {
+        federateUnlisted: true
```
Comment (Chocobozzz) [inline client/src/app/core/server/server.service.ts:63]: Do you want admins to update this setting on the web interface? If no, you can remove this

File: server/controllers/api/config.ts [hunk 5]
```diff
@@ -348,6 +353,11 @@ function customConfig (): CustomConfig {
     theme: {
       default: CONFIG.THEME.DEFAULT
     },
+    federation: {
+      videos: {
+        federateUnlisted: CONFIG.FEDERATION.VIDEOS.FEDERATE_UNLISTED
```
Comment (Chocobozzz) [inline server/controllers/api/config.ts:358]: Do you want admins to update this setting on the web interface? If no, you can remove this

File: server/initializers/config.ts [hunk 6]
```diff
@@ -154,6 +154,11 @@ const CONFIG = {
       URL: config.get<string>('plugins.index.url')
     }
   },
+  FEDERATION: {
+    VIDEOS: {
+      get FEDERATE_UNLISTED () { return config.get<boolean>('federation.videos.federate_unlisted') }
```
Comment (Chocobozzz) [inline server/initializers/config.ts:159]: Do you want admins to update this setting on the web interface? If no, don't use a getter

File: server/tests/api/check-params/config.ts [hunk 7]
```diff
@@ -47,6 +47,11 @@ describe('Test config API validators', function () {
         css: 'body { background-color: red; }'
       }
     },
+    federation: {
+      videos: {
+        federateUnlisted: false
```
Comment (Chocobozzz) [inline server/tests/api/check-params/config.ts:52]: Do you want admins to update this setting on the web interface? If no, you can remove this

File: server/tests/api/videos/video-privacy.ts [hunk 8]
```diff
@@ -164,6 +175,45 @@ describe('Test video privacy', function () {
     }
   })
 
+  it('Should upload a non-federating unlisted video to server 2', async function () {
+    this.timeout(30000)
+
+    customConfig.federation.videos.federateUnlisted = false
+    await updateCustomConfig(servers[1].url, servers[1].accessToken, customConfig)
```
Comment (Chocobozzz) [inline server/tests/api/videos/video-privacy.ts:182]: If the setting cannot be updated using the web interface, take example on https://github.com/Chocobozzz/PeerTube/blob/develop/server/tests/api/redundancy/redundancy-constraints.ts#L124 to update the server config

File: client/src/app/+admin/config/edit-custom-config/edit-custom-config.component.ts (hunk 3 above)
Comment (Tak) [reply client/src/app/+admin/config/edit-custom-config/edit-custom-config.component.ts:142]: Is there any reason not to allow this? I added it to `production.yaml.example` as well, based on your earlier comment.

Comment (Chocobozzz) [reply client/src/app/+admin/config/edit-custom-config/edit-custom-config.component.ts:142]: > Is there any reason not to allow this?

It subjective, but it's not a setting admins will update frequently. So I don't think it's very useful to put this in the admin interface. But you're the one who does, so you're the one who decides :)

Comment (Tak) [reply client/src/app/+admin/config/edit-custom-config/edit-custom-config.component.ts:142]: Hm, I guess the main benefit would be discoverability.
I'm open to whatever you think - if you feel like it doesn't belong in the web interface, I'm willing to do the rework.

Comment (Chocobozzz) [reply client/src/app/+admin/config/edit-custom-config/edit-custom-config.component.ts:142]: Then let's move this out of the web interface. Admins should read the config' file when they install peertube, and we'll mention this in the changelog. You could also mention this setting in https://docs.joinpeertube.org/#/admin-following-instances?id=being-followed-by-an-instance

Comment (Tak) [reply client/src/app/+admin/config/edit-custom-config/edit-custom-config.component.ts]: Ok, done

**Comment by Chocobozzz** [comment]
Thanks @Tak!
